### PR TITLE
Enhancements for `extensions` webhook lib

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area open-source
/area ops-productivity
/area security
/kind enhancement

**What this PR does / why we need it**:
This PR enhances the the auto registration and certificate management in the `extensions` library. Concretely, it allows to handle and differentiate between `MutatingWebhookConfigurations` and `ValidatingWebhookConfigurations`. New options are added to webhook generation functions as well.
The goal is to allow extension admission webhooks (e.g. [admission-aws](https://github.com/gardener/gardener-extension-provider-aws/)) to make use of these functionalities which makes an external management of CA and server certificates obsolete.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
This [WIP branch](https://github.com/timuthy/gardener-extension-provider-aws/tree/enhancement.admission-controller-cert) serves as a reference and testing base for the introduced changes (back ported to `gardener v1.80.3`).
PR remains in draft state until fully qualified for admission and extension controllers.

**Release note**:

```feature dependency
Webhook registration `webhookcmd.NewAddToManagerOptions` can now be used for admission controllers performing validation and mutation in the Garden cluster. This option automatically creates and maintains required `{Mutating,Validating}WebhookConfiguration` objects as well as comes with an automated management for CA and server certificates.
```
```breaking dependency
The `webhookcmd.NewAddToManagerSimpleOptions` function was removed, please use `webhookcmd.NewAddToManagerOptions` instead.
```
```breaking dependency
The `extensionswebhook.New` forbids to pass `mutators` and `validators` at the same time. Please use separate webhooks for validating and mutating actions if required.
```
```noteworthy developer
The extension webhook registration does now differentiate between mutating and validating actions and creates matching `ValidatingWebhookConfigration` or `MutatingWebhookConfiguration` objects. Earlier, only `MutatingWebhookConfiguration`s were created.
```

### CODE REVIEW

File: extensions/pkg/webhook/webhook.go [hunk 1]
```diff
@@ -59,24 +64,34 @@ type Args struct {
...
+		objTypes = append(objTypes, objs...)
+		if actionType != ActionMutating {
+			actionType = ActionValidating
+		}
 	}
 
 	for mut, objs := range args.Mutators {
 		builder.WithMutator(mut, objs...)
+		objTypes = append(objTypes, objs...)
+		actionType = ActionMutating
```
Comment (acumino) [inline extensions/pkg/webhook/webhook.go]: `actionType` would always be set to `ActionMutating` if the length of `args.Mutator` > 0 regardless of the length of  `args.Validator`. Is this expected?

File: extensions/pkg/controller/controlplane/genericactuator/actuator.go [hunk 2]
```diff
@@ -243,12 +243,12 @@ func (a *actuator) reconcileControlPlane(
 ) {
 	if a.atomicShootWebhookConfig != nil {
 		value := a.atomicShootWebhookConfig.Load()
-		webhookConfig, ok := value.(*admissionregistrationv1.MutatingWebhookConfiguration)
+		webhookConfig, ok := value.(*webhook.Configs)
 		if !ok {
 			return false, fmt.Errorf("expected *admissionregistrationv1.MutatingWebhookConfiguration, got %T", value)
```
Comment (acumino) [inline extensions/pkg/controller/controlplane/genericactuator/actuator.go]:
```suggestion
			return false, fmt.Errorf("expected *webhook.Configs, got %T", value)
```

**Review by acumino — COMMENTED** [review]
Had a look with @Kostov6 and @dimitar-kostadinov, In general, the PR looks fine only two small comments, will continue the review later.

/assign @acumino @Kostov6 @dimitar-kostadinov

File: extensions/pkg/webhook/webhook.go (hunk 1 above)
Comment (timuthy) [reply extensions/pkg/webhook/webhook.go]: Yes this is expected. Theoretically, you can mix mutators and validators via the builder but the result will still be a single webhook. As soon as one mutator is involved you need a `MutatingwebhookConfiguration`.

**Comment by timuthy** [comment]
Update:
- First test runs with `gardener-extension-admission-aws` were successful.
- The [certificate reconciler](https://github.com/gardener/gardener/blob/master/extensions/pkg/webhook/certificates/reconciler.go) does not differentiate between source and target clusters and thus CA and server certificate `Secret`s are created in the `virtual-garden` cluster. I'm currently checking if we can switch having `{Mutating,Validating}WebhookConfiguration`s and certificate `Secret`s in different clusters.

File: extensions/pkg/webhook/webhook.go [hunk 3]
```diff
@@ -59,24 +64,34 @@ type Args struct {
...
+		logger  = log.Log.WithName(args.Name).WithValues("provider", args.Provider)
+		builder = NewBuilder(mgr, logger)
+	)
 
 	for val, objs := range args.Validators {
 		builder.WithValidator(val, objs...)
+		objTypes = append(objTypes, objs...)
+		if actionType != ActionMutating {
+			actionType = ActionValidating
+		}
```
Comment (Kostov6) [inline extensions/pkg/webhook/webhook.go:94]: What is the meaning of this if statement? Isn't its condition always true that `actionType != ActionMutating` because at this point `actionType` is an unassigned local variable with value ""

**Comment by timuthy** [comment]
PR is ready for review 🔍

**Comment by timuthy** [comment]
@rfranzke it'd to especially having a second look on https://github.com/gardener/gardener/pull/8725/commits/af45cebd6d967ef41f4a44115f61fac4df6f2b16

File: extensions/pkg/webhook/certificates/reconciler.go [hunk 4]
```diff
@@ -80,23 +79,26 @@ type reconciler struct {
...
+// Then it adds the reconciler to the given manager in order to periodically regenerate the webhook secrets.
+// An 'sourceCluster' can be optionally passed to let the certificate secrets be managed in a different cluster.
+func (r *reconciler) AddToManager(ctx context.Context, mgr manager.Manager, sourceCluster cluster.Cluster) error {
 	r.client = mgr.GetClient()
+	r.sourceClient = mgr.GetClient()
+	apiReader := mgr.GetAPIReader()
+
+	if sourceCluster != nil {
+		r.sourceClient = sourceCluster.GetClient()
+		apiReader = sourceCluster.GetClient()
```
Comment (acumino) [inline extensions/pkg/webhook/certificates/reconciler.go]:
```suggestion
		apiReader = sourceCluster.GetAPIReader()
```

File: extensions/pkg/webhook/certificates/reconciler.go [hunk 5]
```diff
@@ -80,23 +79,26 @@ type reconciler struct {
...
+	}
 
-	present, err := isWebhookServerSecretPresent(ctx, mgr.GetAPIReader(), mgr.GetScheme(), r.ServerSecretName, r.Namespace, r.Identity)
+	present, err := isWebhookServerSecretPresent(ctx, apiReader, mgr.GetScheme(), r.ServerSecretName, r.Namespace, r.Identity)
```
Comment (acumino) [inline extensions/pkg/webhook/certificates/reconciler.go]: Scheme passed here should be taken from `sourceCluster.GetScheme()` if `sourceCluster` is not nil.

File: extensions/pkg/webhook/certificates/reconciler.go [hunk 6]
```diff
@@ -175,29 +182,35 @@ func (r *reconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconc
...
-		// ControlPlane actuator
-		if err := extensionswebhook.InjectCABundleIntoWebhookConfig(r.ShootWebhookConfig, caBundleSecret.Data[secretsutils.DataKeyCertificateBundle]); err != nil {
-			return reconcile.Result{}, err
+	if r.ShootWebhookConfigs != nil && r.ShootWebhookConfigs.HasWebhookConfig() {
+		for _, shootWebhookConfig := range r.ShootWebhookConfigs.GetWebhookConfigs() {
+			// update shoot webhook config object (in memory) with the freshly created CA bundle which is also used by the
+			// ControlPlane actuator
+			if err := extensionswebhook.InjectCABundleIntoWebhookConfig(shootWebhookConfig, caBundleSecret.Data[secretsutils.DataKeyCertificateBundle]); err != nil {
+				return reconcile.Result{}, err
+			}
```
Comment (acumino) [inline extensions/pkg/webhook/certificates/reconciler.go:200]: Is new caBundle really getting injected here? `r.ShootWebhookConfigs.GetWebhookConfigs()` returns a new slice for webhooks in `ShootWebhookConfigs` and here `shootWebhookConfig` is updated which is of new slice returned from `r.ShootWebhookConfigs.GetWebhookConfigs()` and not `r.ShootWebhookConfigs`.

File: extensions/pkg/webhook/registration.go [hunk 7]
```diff
@@ -57,6 +58,41 @@ func PrefixedName(componentName string) string {
...
+	}
+	return &deepCopy
+}
+
+// HasWebhookConfig returns true if 'Configs' contains at least one webhook configuration.
+func (c *Configs) HasWebhookConfig() bool {
+	return c.MutatingWebhookConfig != nil || c.ValidatingWebhookConfig != nil
+}
+
 // BuildWebhookConfigs builds MutatingWebhookConfiguration objects for seed and shoots from the given webhooks slice.
```
Comment (acumino) [inline extensions/pkg/webhook/registration.go]:
```suggestion
// BuildWebhookConfigs builds webhook.Configs for seed and shoot from the given webhooks slice.
```

**Review by acumino — CHANGES_REQUESTED** [review]

File: extensions/pkg/webhook/webhook.go [hunk 8]
```diff
@@ -90,7 +109,13 @@ func New(mgr manager.Manager, args Args) (*Webhook, error) {
...
-		Path:    args.Path,
-		Webhook: &admission.Webhook{Handler: handler, RecoverPanic: true},
+		Name:           args.Name,
+		Action:         actionType,
+		Selector:       args.Selector,
+		ObjectSelector: args.ObjectSelector,
+		Path:           args.Path,
+		Target:         args.Target,
+		Webhook:        &admission.Webhook{Handler: handler, RecoverPanic: true},
+		Types:          objTypes,
```
Comment (acumino) [inline extensions/pkg/webhook/webhook.go]: Any reason provider is not passed?
```suggestion
		Provider:       args.Provider,
		Name:           args.Name,
		Action:         actionType,
		Selector:       args.Selector,
		ObjectSelector: args.ObjectSelector,
		Path:           args.Path,
		Target:         args.Target,
		Webhook:        &admission.Webhook{Handler: handler, RecoverPanic: true},
		Types:          objTypes,
```

**Comment by rfranzke** [comment]
I will try to take a look tomorrow.

File: extensions/pkg/webhook/certificates/reconciler.go (hunk 6 above)
Comment (timuthy) [reply extensions/pkg/webhook/certificates/reconciler.go:200]: Hm, they all share the same references (for validating and mutating webhook object) - no matter what slice. Am I missing something?

File: extensions/pkg/webhook/certificates/reconciler.go (hunk 5 above)
Comment (timuthy) [reply extensions/pkg/webhook/certificates/reconciler.go]: Nice catch 👍

File: extensions/pkg/webhook/certificates/reconciler.go (hunk 6 above)
Comment (acumino) [reply extensions/pkg/webhook/certificates/reconciler.go:200]: Just wanted to double-confirm. 😃

File: extensions/pkg/webhook/registration.go [hunk 9]
```diff
@@ -66,73 +84,70 @@ func BuildWebhookConfigs(
...
+	}
+
 	for _, webhook := range webhooks {
 		var rules []admissionregistrationv1.RuleWithOperations
 
 		for _, t := range webhook.Types {
 			rule, err := buildRule(c, t)
 			if err != nil {
-				return nil, nil, err
+				return seedWebhookConfig, nil, err
```
Comment (rfranzke) [inline extensions/pkg/webhook/registration.go]: Why is this needed?

File: extensions/pkg/webhook/registration.go [hunk 10]
```diff
@@ -66,73 +84,70 @@ func BuildWebhookConfigs(
...
-	// if all webhooks for one target are removed in a new version, extensions need to explicitly delete the respective
-	// webhook config
-	if len(seedWebhooks) > 0 {
-		seedWebhookConfig = &admissionregistrationv1.MutatingWebhookConfiguration{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:   NamePrefix + providerName,
-				Labels: map[string]string{v1beta1constants.LabelExcludeWebhookFromRemediation: "true"},
-			},
-			Webhooks: seedWebhooks,
+			return seedWebhookConfig, nil, fmt.Errorf("invalid webhook target: %s", webhook.Target)
```
Comment (rfranzke) [inline extensions/pkg/webhook/registration.go]: same question here

File: extensions/pkg/webhook/webhook.go (hunk 1 above)
Comment (rfranzke) [reply extensions/pkg/webhook/webhook.go]: Should we rather add some validation in the beginning of this function to prevent mixing validators and mutators? It's a bit confusing overall, and one shouldn't do this, or is there a scenario where you would?

File: extensions/pkg/webhook/registration.go [hunk 11]
```diff
@@ -86,15 +103,13 @@ func BuildWebhookConfigs(
 	caBundle []byte,
 ) (
 	seedWebhookConfig Configs,
-	shootWebhookConfig *admissionregistrationv1.MutatingWebhookConfiguration,
+	shootWebhookConfig Configs,
```
Comment (rfranzke) [inline extensions/pkg/webhook/registration.go]: seedWebhookConfig*s* and shootWebhookConfig*s* :)

File: extensions/pkg/webhook/registration.go [hunk 12]
```diff
@@ -368,6 +371,7 @@ func BuildClientConfigFor(webhookPath string, namespace, componentName string, s
 
 func createAndAddToWebhookConfig(
 	webhookConfigs *Configs,
+	nameFunc func() string,
```
Comment (rfranzke) [inline extensions/pkg/webhook/registration.go]: Why is a `func` needed? Can't we use `name string` directly?

**Review by rfranzke — COMMENTED** [review]
Nice PR, quite a ride navigating in this `webhook` package... :)

File: extensions/pkg/webhook/registration.go (hunk 9 above)
Comment (timuthy) [reply extensions/pkg/webhook/registration.go]: Hm, it has to return something?! Earlier it was a pointer, hence return `nil`, but this has been changed because it simplifies the overall handling a bit.

Comment (rfranzke) [reply extensions/pkg/webhook/registration.go]: Aha, I missed that it's no longer a pointer now, so `nil` cannot be returned. Thanks!

**Comment by timuthy** [comment]
@rfranzke I don't know why answering your [comment](https://github.com/gardener/gardener/pull/8725#discussion_r1399122445) is not possible, so I do it here:

> Should we rather add some validation in the beginning of this function to prevent mixing validators and mutators? It's a bit confusing overall, and one shouldn't do this, or is there a scenario where you would?

I didn't add validation for two reasons:
- It's not backwards compatible
- I don't see a major issue mixing mutating and validating functions. Earlier, we always used `MutatingWebhookConfiguration`s and it didn't pose any problems.

**Comment by rfranzke** [comment]
> @rfranzke I don't know why answering your [comment](https://github.com/gardener/gardener/pull/8725#discussion_r1399122445) is not possible, so I do it here:
> 
> > Should we rather add some validation in the beginning of this function to prevent mixing validators and mutators? It's a bit confusing overall, and one shouldn't do this, or is there a scenario where you would?
> 
> I didn't add validation for two reasons:
> 
> * It's not backwards compatible

OK. Should we add a release note and a TODO statement to add it, eventually? I guess it would make sense to force users to clearly separate their webhooks instead of mixing them together?

> * I don't see a major issue mixing mutating and validating functions. Earlier, we always used `MutatingWebhookConfiguration`s and it didn't pose any problems.

Yes, this works technically, but I assume it can be quite confusing, especially during operations or debugging activities, don't you think?

**Comment by acumino** [comment]
> I don't see a major issue mixing mutating and validating functions. Earlier, we always used MutatingWebhookConfigurations and it didn't pose any problems.

Mutating webhooks works before validating webhooks and there can be a scenario where some validation passes after the changes performed by mutating webhooks.
Would not mixing mutation and validation can cause issues in that case?

File: extensions/pkg/webhook/webhook_test.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	})
+})
+
+type fakeManager struct {
+	manager.Manager
+}
+
+func (f *fakeManager) GetScheme() *runtime.Scheme {
+	return kubernetesscheme.Scheme
+}
```
Comment (rfranzke) [inline extensions/pkg/webhook/webhook_test.go]: Nit: Maybe use https://github.com/gardener/gardener/blob/master/pkg/utils/test/manager.go?

**Comment by timuthy** [comment]
> Mutating webhooks works before validating webhooks and there can be a scenario where some validation passes after the changes performed by mutating webhooks.
Would not mixing mutation and validation can cause issues in that case?

Although technically true, it'd be odd and wrong to rely on this behavior. At least no component in the `github.com/gardener` org is affected by this change.

**Review by rfranzke — COMMENTED** [review]
/lgtm

**Review by acumino — COMMENTED** [review]
/approve
